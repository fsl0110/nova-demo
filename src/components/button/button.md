### A Button-Component

#### Button with only text

```jsx
<Button style={{ background: "pink" }} className="error__button">
  Button text
</Button>
```

#### Button with only icon in size xl

```jsx
<Button className="search__button" iconName="search" iconSize="xl" />
```

#### Button disabled with only icon in size xl

```jsx
<Button
  className="search__button"
  iconName="search"
  iconSize="xl"
  isLoading={true}
/>
```

#### Button with text left and icon right

```jsx
<Button
  style={{ background: "pink" }}
  className="error__button"
  iconName="refresh"
  iconAlign="after"
>
  Button Text
</Button>
```

#### Button with text right and icon left

```jsx
<Button
  style={{ background: "pink" }}
  className="error__button"
  iconName="refresh"
  iconAlign="before"
>
  Button Text
</Button>
```

#### Button disabled while loading

```jsx
<Button
  style={{ background: "pink" }}
  className="error__button"
  iconName="refresh"
  iconAlign="after"
  isLoading={true}
>
  Button Text
</Button>
```

#### Button with spinning icon while loading

```jsx
<Button
  style={{ background: "pink" }}
  className="error__button"
  iconName="refresh"
  iconAlign="after"
  isLoading={true}
  isLoadingSpin={true}
>
  Button Text
</Button>
```
