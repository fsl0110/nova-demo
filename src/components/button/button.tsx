import React, { PureComponent, ReactNode, HTMLAttributes } from "react";
import classNames from "classnames";
import { Icon } from "..";
import { IconNameProps, IconSizeProps, IconRotateProps } from "../icon/icon";

export type AlignProps = "before" | "after";

export interface Props extends HTMLAttributes<HTMLButtonElement> {
  children?: ReactNode | ReactNode[];
  /** IconName of the icon used before or after the text. */
  iconName?: IconNameProps;
  /** Alignment of the icon. Default is before. */
  iconAlign?: AlignProps;
  /** Size of the icon. */
  iconSize?: IconSizeProps;
  /** While isLoading is true the button is disabled */
  isLoading?: boolean;
  /** Should the icon spinning while isLoading is active? */
  isLoadingSpin?: boolean;
}

export class Button extends PureComponent<Props> {
  render() {
    const {
      iconName,
      iconSize,
      iconAlign,
      isLoading,
      isLoadingSpin,
      children,
      className,
      ...rest
    } = this.props;

    if (!children && !iconName) {
      return null;
    }

    const rootClasses = classNames("fs-btn", className, {
      [`fs-btn--${iconAlign}`]: !!iconAlign
    });

    return (
      <button disabled={isLoading} className={rootClasses} {...rest}>
        <Icon
          name={iconName}
          spin={isLoadingSpin && isLoading}
          size={iconSize}
        />
        {children && <span>{children}</span>}
      </button>
    );
  }
}

export default Button;
