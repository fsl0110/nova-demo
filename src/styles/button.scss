.fs-btn {
  outline: none;
  background: none;
  border: inherit;
  display: flex;
  justify-content: space-between;
  /*   align-items: center; */
}

.fs-btn:focus {
  outline: none;
}

.fs-btn--after {
  flex-direction: row-reverse;
}

.fs-btn--before {
  flex-direction: row;
}
