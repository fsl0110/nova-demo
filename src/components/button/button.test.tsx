import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import { Button, Props } from "./button";

const testCases: Array<[string, Props]> = [
  [
    "Button with only text",
    {
      children: "Button text"
    }
  ],
  [
    "Button with only icon in size xl",
    {
      iconName: "refresh",
      iconSize: "xl"
    }
  ],
  [
    "Button with text left and icon right",
    {
      children: "Button text",
      iconName: "refresh",
      iconAlign: "after"
    }
  ],
  [
    "Button with text right and icon left",
    {
      children: "Button text",
      iconName: "refresh",
      iconAlign: "before"
    }
  ],
  [
    "Button disabled while loading",
    {
      children: "Button text",
      iconName: "refresh",
      iconAlign: "before",
      isLoading: true
    }
  ],
  [
    "Button with spinning icon while loading",
    {
      children: "Button text",
      iconName: "refresh",
      iconAlign: "after",
      isLoading: true,
      isLoadingSpin: true
    }
  ]
];

describe("<Button>", () => {
  testCases.forEach(([item, options]) => {
    it(item, () => {
      const mock = shallow(<Button {...options}>Some text</Button>);
      expect(toJson(mock)).toMatchSnapshot();
    });
  });
});
