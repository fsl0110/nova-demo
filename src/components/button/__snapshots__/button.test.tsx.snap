// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> Button disabled while isLoading 1`] = `
<button
  className="fs-btn fs-btn--before"
  disabled={true}
>
  <Icon
    name="refresh"
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button disabled while loading 1`] = `
<button
  className="fs-btn fs-btn--before"
  disabled={true}
>
  <Icon
    name="refresh"
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button with a text left an icon right 1`] = `
<button
  className="fs-btn fs-btn--after"
>
  <Icon
    name="refresh"
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button with a text right an icon left 1`] = `
<button
  className="fs-btn fs-btn--before"
>
  <Icon
    name="refresh"
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button with only a text 1`] = `
<button
  className="fs-btn"
>
  <Icon
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button with only an icon in size xl 1`] = `
<button
  className="fs-btn"
>
  <Icon
    name="refresh"
    size="xl"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button with only icon in size xl 1`] = `
<button
  className="fs-btn"
>
  <Icon
    name="refresh"
    size="xl"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button with only text 1`] = `
<button
  className="fs-btn"
>
  <Icon
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button with spinning icon while isLoading 1`] = `
<button
  className="fs-btn fs-btn--after"
  disabled={true}
>
  <Icon
    name="refresh"
    size="s"
    spin={true}
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button with spinning icon while loading 1`] = `
<button
  className="fs-btn fs-btn--after"
  disabled={true}
>
  <Icon
    name="refresh"
    size="s"
    spin={true}
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button with text left and icon right 1`] = `
<button
  className="fs-btn fs-btn--after"
>
  <Icon
    name="refresh"
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> Button with text right and icon left 1`] = `
<button
  className="fs-btn fs-btn--before"
>
  <Icon
    name="refresh"
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> renders a button disabled while isLoading 1`] = `
<button
  className="fs-btn fs-btn--before"
  disabled={true}
>
  <Icon
    name="refresh"
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> renders a button with a text left an icon right 1`] = `
<button
  className="fs-btn fs-btn--after"
>
  <Icon
    name="refresh"
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> renders a button with a text right an icon left 1`] = `
<button
  className="fs-btn fs-btn--before"
>
  <Icon
    name="refresh"
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> renders a button with only a text 1`] = `
<button
  className="fs-btn"
>
  <Icon
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> renders a button with only an icon in size xl 1`] = `
<button
  className="fs-btn"
>
  <Icon
    name="refresh"
    size="xl"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> renders a button with spinning icon 1`] = `
<button
  className="fs-btn fs-btn--after"
  disabled={true}
>
  <Icon
    name="refresh"
    size="s"
    spin={true}
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> renders a disabled button 1`] = `
<button
  className="fs-btn fs-btn--before"
  disabled={true}
>
  <Icon
    name="refresh"
    size="s"
  />
  <span>
    Some text
  </span>
</button>
`;

exports[`<Button> renders a spinning isLoading button 1`] = `
<button
  className="fs-btn fs-btn--after"
  disabled={true}
>
  <Icon
    name="refresh"
    size="s"
    spin={true}
  />
  <span>
    Some text
  </span>
</button>
`;
